### ISSUE TITLE
1.5.3 `UsePrivacyMode=y` causes random startup failures for Firefox

### ISSUE THREAD

**Issue by ksmustang** [sandboxie-plus/Sandboxie#2588]
**Updated Jan 10, 2023
Further regression testing reveals the problem was introduced in version 1.5.3. Probably by "fixed UsePrivacyMode=y compatibility with Windows 7 #2423"**

**SBIE+ 1.5.3** with Data Protection enabled (`UsePrivacyMode=y`) causes Firefox to randomly have problems starting up.  Problems do not occur with `UsePrivacyMode=n`. Versions 1.6.2, 1.6.3, and 1.6.4 all exhibit this behavior. The trace files I've attached were made on 1.6.3.

The problem appears to be completely random. Most of the time Firefox starts normally, but 1 out of every 10 to 20 startups will fail in one or the following ways:
- Firefox processes start then stop (no Firefox window ever appears)
- Firefox processes start then hang (no Firefox window ever appears, the processes must be stopped manually)
- Firefox processes start, the Firefox window appears, but Firefox displays nothing from the web page. The window title shows Firefox is not responding. Firefox must be force closed.
- Firefox processes start, the Firefox window appears and is partially filled, but is non-responsive. After a delay, one or more of the following window appears:
![FF_Unexpected_Parameters](https://user-images.githubusercontent.com/37599506/210664345-0f363549-993c-41b0-b35a-b89520795a4c.jpg)
Firefox closes after clicking ‘OK’ on the window(s) in this failure case.

It would seem a timing issue of some sort between Sandboxie and Firefox is responsible for the random startup failures with `UsePrivacyMode=y`. This problem did not occur with 1.5.1.

The Dropbox link below contains a zip file with the following folders:

- Firefox_ini: contains the Firefox `Installs.ini` and `Profiles.ini` files (located in `%AppData%\Mozilla\Firefox`) used for testing
- Firefox_Profile: contains a zip file for the Firefox profile folder corresponding to the default profile in the `Profiles.ini`file: `6ymsxbh8.default-release`.  (Located in `%AppData%\Mozilla\Firefox\Profiles`) 
- Sandbox_ini: contains the `sandboxie.ini`file used for testing
- Trace_files: contains three Sandboxie trace files: (1) a trace when Firefox starts normally, (2) a trace when Firefox hangs with no window, (3) a trace when Firefox processes start, the Firefox window appears and is partially filled but is non-responsive and the above Firefox error window is displayed.

I've tested with Firefox 106.0.2 and 108.0.1 (64-bit) and the problems occur with both versions.

### How often did you encounter it so far?

The problem is random. When repeatedly starting Firefox, it will sometimes fail to start properly

### Affected program

Firefox 108.0.1 64-bit

### Download link

Nor relevant

### Where is the program located?

The program is installed only outside the sandbox.

### Expected behavior

Firefox starts normally every time.

### What is your Windows edition and version?

Windows 10 Pro 21H2 64-bit

### In which Windows account you have this problem?

A local or Microsoft account without special changes.

### Please mention any installed security software

Standard Windows Security

### What version of Sandboxie are you running?

Sandboxie Plus 1.6.3 64-bit

### Is it a new installation of Sandboxie?

I just updated Sandboxie from a previous version (to be specified).

### Is it a regression?

Does not occur in 1.5.1, first observed in 1.6.2

### In which sandbox type you have this problem?

In a sandbox with data protection (blue sandbox icon).

### Can you reproduce this problem on a new empty sandbox?

I can confirm it also on a new empty sandbox.

### Did you previously enable some security policy settings outside Sandboxie?

No

### Trace log

https://www.dropbox.com/s/komg58oh7sp3ocs/FireFoxRandomStartupFailWithDataProtection.zip?dl=0

### Sandboxie.ini configuration

```shell
#
# Sandboxie configuration file
#

[GlobalSettings]
DefaultBox=DefaultBox
FileRootPath=\??\%SystemDrive%\Sandbox\%USER%\%SANDBOX%
KeyRootPath=\REGISTRY\USER\Sandbox_%USER%_%SANDBOX%
IpcRootPath=\Sandbox\%USER%\%SANDBOX%\Session_%SESSION%

[UserSettings_052E014C]
SbieCtrl_AutoStartAgent=SandMan.exe
SbieCtrl_EnableAutoStart=y

[DefaultBox]
Enabled=y
BlockNetworkFiles=y
BorderColor=#0000ff,ttl,6
Template=OpenBluetooth
Template=SkipHook
Template=FileCopy
Template=qWave
Template=BlockPorts
Template=LingerPrograms
Template=AutoRecoverIgnore
ConfigLevel=9
ForceProcess=firefox.exe
BoxNameTitle=n
UseSecurityMode=n
CopyLimitKb=81920
AutoDelete=y

UsePrivacyMode=y
NormalFilePath=%AppData%\Mozilla\

FileTrace=*
PipeTrace=*
KeyTrace=*
IpcTrace=*
GuiTrace=*
ClsidTrace=*
DebugTrace=y
```

**Comment by DavidXanatos** [comment]
is 1.6.1 working ok then?

**Comment by ksmustang** [comment]
Don't know, sorry. I'm currently running 1.5.1 and 1.6.2 was the current version when I ran into the issue. I can test 1.6.1 for you if that would be helpful.

**Comment by isaak654** [comment]
In that case, the commit to look at would be https://github.com/sandboxie-plus/Sandboxie/commit/48504e1db14522d7123b02322995fd6d9d4f99ad

**Comment by DavidXanatos** [comment]
iirc the change was in how the box is initialized, 
could you please try to inisialize a sandbox on the last working build and then keep using it without deleting its content on the newer build,
and let me know if than the issue is gone

**Comment by ksmustang** [comment]
@DavidXanatos: The sandbox contents are auto-deleted, so that test doesn't seem to apply.

I have done a lot of additional testing. While the problem definitely began with 1.5.3, I may have been premature to believe it was due to the fix for #2482.

I used the Sandboxie tab in TaskExplorer (great program!) to see the file and registry entries generated for a sandbox. I could see Sandboxie generates multiple `WriteFilePath=drive` entries when `UsePrivacyMode=y` is present.

So, I ran a test where I changed `UsePrivacyMode=y` to `UsePrivacyMode=n` and added the following entries. The same random startup failures occurred when testing this configuration.

- - `WriteFilePath=C:\`
- - `WriteFilePath=E:\`    (2nd drive on my computer)

I then changed the two `WriteFilePath=` entries to `ClosedFilePath=` and tested. **The random startup problem stopped!**


What I've learned from my testing is that the random startup problem with `UsePrivacyMode=y` appears to occur as a result of the combination of:

- - Programs that initiate a lot of processes on startup (like, Firefox, Edge, Chrome)     -AND-
- - `WriteFilePath=drive`

Again, the problem does not occur every time I start Firefox. But it does happen in 1 out of every 10 to 20 startups. The problem manifests itself in several ways as I described in the original issue report. And if the problem occurs, it occurs while Firefox is initiating the multiple processes it creates on startup.

**Comment by bastik-1001** [comment]
This might be related to how Firefox checks if it is the default browser. This is just a guess, based on the fact, that when I start Firefox without privacy mode, it sees itself as the default browser (which it is), but in privacy mode, it asks if I want to make it the default browser. It fails, when attempting to change the setting.

**Comment by ksmustang** [comment]
The default browser check by Firefox does not appear to be a factor in the random startup failures. And random startup failures also occur for Edge and Chrome, but Firefox is my primary browser.

I've done a lot of testing and the the problem is related to the `WriteFilePath=` Sandboxie generates when using privacy mode. I have been able to eliminate the startup failures by using privacy mode and adding two `ClosedFilePath=` statements:

C:\Users\%USER%\AppData
E:\Internet

I also reported this issue in the Sandboxie forum on Wilders in the v1.8.0 release thread. Refer to the following link for more details:
https://www.wilderssecurity.com/threads/sandboxie-plus-v1-8-0.450696/page-2#post-3135796

**Comment by bastik-1001** [comment]
I did not run Firefox with privacy mode enabled, I just for testing and it not being able to detect that it was the default browser, made me jump to the guess it could be related to it.

While starting and closing Firefox in a secured box with privacy protection (red color box), clearing the boxes content, there were instances at which Firefox failed. I showed up in Sandman, but then got closed, with no error messages of anything nor an entry in the log. Sometimes I got a window, but before doing anything, the process just crashed, again without any error or without something. Even the crash manager of Firefox did not show up, should it have been triggered.

I added `NormalFilePath=C:\Users\%USER%\*` and `WriteFilePath=C:\Users\*` to the config (which I had removed before enabling the privacy mode), what gives the sandbox way more access than the original privacy mode (which is still enabled), but since I added these, there are no crashes of Firefox. (It's still more restrictive than how I used to run it.)

Thank you for your testing, this should help a lot to fix the underlying issue.

**Comment by DavidXanatos** [comment]
Could you please try narrowing in which folders in the user profile are actually required?

**Comment by ksmustang** [comment]
I will do that. It may be a day or two before I can send results.On Mar 27, 2023, at 2:07 AM, DavidXanatos ***@***.***> wrote:﻿
Could you please try narrowing in which folders in the user profile are actually required?

**Comment by ksmustang** [comment]
> Could you please try narrowing in which folders in the user profile are actually required?

I have performed literally hundreds of tests with Firefox, Edge and Chrome to try to narrow down what is needed to prevent random startup and/or lockup failures. The frequency of startup failures varies, making it necessary to run many tests with every setting change for each of the browsers.

As a result of the testing, I found I need to expand the scope of the `ClosedFilePath` rather than narrow it. I now use `ClosedFilePath=C:\Users\%USER%` for Firefox, Edge and Chrome to eliminate random startup and/or lockup failures.

These failures happen on my physical laptop as well as in a VM with far fewer applications installed.

Don't know why I'm experiencing these problems and others don't seem to be affected. I know the frequency of the problems varies but I cannot identify the factors causing it. Perhaps others are experiencing problems, but at a frequency low enough that they just deal with it.

**Comment by offhub** [comment]
Are there any spaces in the Windows account name?

**Comment by bastik-1001** [comment]
> Are there any spaces in the Windows account name?

Not in mine, but it contains `_` (just to give feedback, even though this is my report and I can't give as much insight.)

**Comment by offhub** [comment]
@ksmustang How do you run browsers when you close that path?

**Comment by offhub** [comment]
@DavidXanatos 

https://user-images.githubusercontent.com/6871698/230907775-cc111c6f-d3ca-49d0-bc98-3e36b2a7bab1.mp4

**Comment by ksmustang** [comment]
> Are there any spaces in the Windows account name?

No. My account name is 3 alphabetic characters



> @ksmustang How do you run browsers when you close that path?

By adding the appropriate `NormalFilePath` and `OpenFilePath` statements to the sandbox in order to allow the browser to access the folders/files it needs in the user profile.

Keep in mind that a data protected sandbox causes Sandboxie to automatically generate some `WriteFilePath` statements that lock down access to the drives on the system. It also automatically generates some `NormalFilePath` statements to allow access  to `C:\Program FIles`, `C:\Program Files (x86)`, and `C:\Windows`. It is then up to you to determine what other folders/files a program needed to access in order to run.

What I found is that something about the auto-generated `WriteFilePath` statements causes intermittent startup/lockup problems for some programs and that adding `ClosedFilePath` for certain folders to the sandbox configuration seems to eliminate the problems. The challenge is to determine which folders are causing the problems. In my testing, it seems that the user profile folders seem to be the 'culprit'.

**Comment by DavidXanatos** [comment]
Please try if removing 
```
NormalFilePath=%AppData%\Mozilla\
```

remedies the issue?

**Comment by DavidXanatos** [comment]
Please try the latest 1.9.0 pre release build with this 1.9.1 dll's
[SbieDll1.9.1.zip](https://github.com/sandboxie-plus/Sandboxie/files/11284044/SbieDll1.9.1.zip)

**Comment by ksmustang** [comment]
> Please try the latest 1.9.0 pre release build with this 1.9.1 dll's [SbieDll1.9.1.zip](https://github.com/sandboxie-plus/Sandboxie/files/11284044/SbieDll1.9.1.zip)

@DavidXanatos 

Pre-release 1.9.0  with SbieDll.dll for 1.9.1 fixes the startup and other problems I've encountered since 1.5.3 for firefox, chrome and edge! I tested each of the three browsers 100+ times; there were no startup or other failures in the total of 300+ tests. 

As I noted when I opened the issue, 1.5.3 introduced a problem that caused the browsers to have random startup or other failures, usually within the first 20 tests. So I consider this problem to be fixed with these changes.

I'm curious what you found to correct the problem?

Many thanks for your outstanding support and for the tremendous improvements you've made to Sandboxie!
